astropy/units/format/console.py

```diff
@@ -71,4 +71,4 @@ def to_string(
     ) -> str:
         # Change default of fraction to False, i.e., we typeset
         # without a fraction by default.
-        return super().to_string(unit, fraction=fraction)
+        return cls._to_string(unit, fraction=fraction)
```

Comment:
The big problem here was that `Console.to_string()` accepts `fraction="multiline"` and can pass that to `Base.to_string()`, but `Base.to_string()` does not accept `fraction="multiline"`. The reason this works just fine at runtime anyways is that the code gets a dictionary from the `_fraction_formatters` class property and uses the `fraction` argument to look up the appropriate method to call from there, but such code is too dynamic for type checkers to understand. The `Base._to_string()` method introduced in 5afb6506c7a0ceb69d27bda14524c7883c72ed52 allows rewriting this in a way that does not cause spurious type errors.

As far as type checking is concerned this solution is not perfect either because type checkers cannot enforce that the annotation of `fraction` matches the keys in `_fraction_formatters`. It is still the responsibility of the developer to ensure that. It would be possible avoid `_fraction_formatters` altogether by providing multiple implementations of `_to_string()` (in `Base` and in `Console`), but that means code duplication. It would also be possible to use [`TypedDict`](https://docs.python.org/3.11/library/typing.html#typing.TypedDict), but I think that would be too much boilerplate for this code. If you have any better ideas then do let me know.

Reply:
I was going to suggest TypedDict as well. The problem is also that this performs no runtime value check.

Reply:
I think the conclusion is that `Base` should accept multiline and just error on it. We had a long discussion about this at the time... I quite dislike the pattern here of not using `super()`, and feel it is being forced because of a bad previous decision - "multiline" should be a *request* to do multi-line if that's possible, not a requirement that breaks code if the format does not support it.

Reply:
> I quite dislike the pattern here of not using `super()`, and feel it is being forced because of a bad previous decision

I agree that the way the `fraction` parameter is used in the current API is very awkward and we might want to change it, but making sure that the new API is friendly towards type checkers requires some care. If you are interested in opening a pull request where we could discuss this in detail then I can drop the changes to the `to_string()` methods here.

Reply:
Yes, I'm interested in doing that! Though I won't have time till next week, as I'm still traveling.